samples/matter/common/src/bridge/ble_connectivity_manager.cpp

```diff
@@ -315,81 +325,149 @@ CHIP_ERROR BLEConnectivityManager::Reconnect()
-					   void *context, bt_uuid *serviceUuid)
+CHIP_ERROR BLEConnectivityManager::Connect(BLEBridgedDeviceProvider *provider)
 {
-	/* Check if device selected to connect is present on the scanned devices list. */
-	if (mScannedDevicesCounter <= index) {
-		return CHIP_ERROR_INVALID_ARGUMENT;
-	}
-
-	/* Verify whether there is a free space for a new device. */
-	if (mCreatedDevicesCounter >= kMaxCreatedDevices) {
-		return CHIP_ERROR_NO_MEMORY;
-	}
-
-	if (!callback || !serviceUuid) {
+	if (!provider) {
 		return CHIP_ERROR_INVALID_ARGUMENT;
 	}
 
 	StopScan();
 
 	bt_conn *conn;
+	bt_le_conn_param *connParams = GetScannedDeviceConnParams(provider->GetBLEBridgedDevice().mAddr);
+
+	if (!connParams) {
+		LOG_ERR("Failed to get conn params");
+		return CHIP_ERROR_INTERNAL;
+	}
 
-	int err = bt_conn_le_create(&mScannedDevices[index].mAddr, create_param, &mScannedDevices[index].mConnParam,
-				    &conn);
+	int err = bt_conn_le_create(&provider->GetBLEBridgedDevice().mAddr, create_param, connParams, &conn);
 
 	if (err) {
 		LOG_ERR("Creating connection failed (err %d)", err);
 		return System::MapErrorZephyr(err);
+	} else {
+		provider->SetConnectionObject(conn);
+	}
+
+	return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BLEConnectivityManager::AddBLEProvider(BLEBridgedDeviceProvider *provider)
+{
+	if (!provider) {
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	if (mConnectedProvidersCounter >= kMaxConnectedDevices) {
+		return CHIP_ERROR_NO_MEMORY;
+	}
+
+	/* Find first free slot to store new providers' address. */
+	for (auto i = 0; i < kMaxConnectedDevices; i++) {
+		if (mConnectedProviders[i] == nullptr) {
+			mConnectedProviders[i] = provider;
+			mConnectedProvidersCounter++;
+			return CHIP_NO_ERROR;
+		}
 	}
 
-	mCreatedDevices[mCreatedDevicesCounter].mAddr = mScannedDevices[index].mAddr;
-	mCreatedDevices[mCreatedDevicesCounter].mFirstConnectionCallback = callback;
-	mCreatedDevices[mCreatedDevicesCounter].mFirstConnectionCallbackContext = context;
-	mCreatedDevices[mCreatedDevicesCounter].mServiceUuid = serviceUuid;
-	mCreatedDevicesCounter++;
+	return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR BLEConnectivityManager::RemoveBLEProvider(bt_addr_le_t address)
+{
+	BLEBridgedDeviceProvider *provider = nullptr;
+
+	/* Find provider's address on the list and remove it. */
+	for (auto i = 0; i < kMaxConnectedDevices; i++) {
+		if (mConnectedProviders[i] == nullptr) {
+			continue;
+		}
+
+		bt_addr_le_t addr = mConnectedProviders[i]->GetBtAddress();
+		if (memcmp(&addr, &address, sizeof(address)) == 0) {
+			provider = mConnectedProviders[i];
+			mConnectedProviders[i] = nullptr;
+			mConnectedProvidersCounter--;
+			break;
+		}
+	}
+
+	if (!provider) {
+		return CHIP_ERROR_NOT_FOUND;
+	}
+
+	if (!provider->GetBLEBridgedDevice().mConn) {
+		return CHIP_ERROR_INTERNAL;
+	}
+
+	/* Release the connection immediately in case of error, as callback will not be called. */
```

Comment:
What does it mean "in case of error". I can see that in case of error (we don't have not provider nor mConn), we return before execution of the following part of this method.

Reply:
Rephrased to "disconnection failed". I mean that we cannot unref the connection if provider or connection object is null. We can do that if for some reason disconnect would fail. Otherwise we may end with not decremented number of connections what may prevent from adding new ones, as the pool will be full.